ouroboros-consensus/test-storage/Test/Ouroboros/Storage/VolatileDB/StateMachine.hs

```diff
@@ -113,161 +142,379 @@ instance CommandNames (At Cmd) where
+transitionImpl :: Model r -> At CmdErr r -> At Resp r -> Model r
 transitionImpl model cmd = eventAfter . lockstep model cmd
 
-preconditionImpl :: Model Symbolic -> At Cmd Symbolic -> Logic
-preconditionImpl m@Model {..} (At cmd) =
+preconditionImpl :: Model Symbolic -> At CmdErr Symbolic -> Logic
+preconditionImpl m@Model {..} (At (CmdErr cmd _)) =
     Not (knownLimitation m (At cmd))
+    .&& case cmd of
+        Corrupt cors ->
+            forall (corruptionFiles cors) (`elem` getDBFiles dbModel)
+            .&& (Boolean $ open dbModel)
+        _ -> Top
+  where
+      corruptionFiles = map snd . NE.toList
 
-postconditionImpl :: Model Concrete -> At Cmd Concrete -> At Resp Concrete -> Logic
-postconditionImpl model cmd resp =
+postconditionImpl :: Model Concrete -> At CmdErr Concrete -> At Resp Concrete -> Logic
+postconditionImpl model cmdErr resp =
     toMock (eventAfter ev) resp .== eventMockResp ev
   where
-    ev = lockstep model cmd resp
+    ev = lockstep model cmdErr resp
 
-generatorImpl :: Model Symbolic -> Maybe (Gen (At Cmd Symbolic))
-generatorImpl Model {..} = Just $ At <$> do
-    sl <- arbitrary
+generatorCmdImpl :: Model Symbolic -> Maybe (Gen (At Cmd Symbolic))
+generatorCmdImpl m@Model {..} =
+    if shouldEnd then Nothing else Just $ do
+    sl <- blockIdgenerator m
     let lastGC = latestGarbaged dbModel
-    ls <- filter (newer lastGC . toSlot) <$> arbitrary
-    frequency
-        [ (3, return $ GetBlock sl)
-        , (3, return $ PutBlock sl)
-        , (1, return $ GarbageCollect sl)
-        , (1, return $ IsOpen)
-        , (1, return $ Close)
-        , (1, return $ ReOpen)
-        , (if null ls then 0 else 1, return $ AskIfMember ls)
+    let dbFiles :: [String] = getDBFiles dbModel
+    ls <- filter (newer lastGC . toSlot) <$> (listOf $ blockIdgenerator m)
+    cmd <- frequency
+        [ (15, return $ GetBlock sl)
+        , (15, return $ PutBlock sl)
+        , (5, return $ GarbageCollect sl)
+        , (5, return $ IsOpen)
+        , (5, return $ Close)
+        -- When the db is Closed, we try to ReOpen it asap.
+        -- This helps minimize TagClosedError and create more
+        -- interesting tests.
+        , (if open dbModel then 1 else 100, return $ ReOpen)
+        , (if null ls then 0 else 3, return $ AskIfMember ls)
+        , (if null dbFiles then 0 else 3, Corrupt <$> generateCorruptions (NE.fromList dbFiles))
         ]
+    return $ At cmd
+
+generatorImpl :: Bool -> Model Symbolic -> Maybe (Gen (At CmdErr Symbolic))
+generatorImpl mkErr m@Model {..} = do
+    genCmd <- generatorCmdImpl m
+    Just $ do
+        At cmd <- genCmd
+        err' <- if noErrorFor cmd then return Nothing
+           else frequency
+                [ (1, return Nothing)
+                , (if mkErr then 1 else 0, Just <$> arbitrary)]
+        let err = erasePutCorruptions err'
+        return $ At $ CmdErr cmd err
+    where
+        eraseCorruptions str = (\(fsErr, _) -> (fsErr, Nothing)) <$> str
```

Comment:
This needs a comment why this is necessary and why this doesn't reduce the power of the test (because you have the `AppendBytes` constructor).